src/cpp/src/lm_encoding.cpp

```diff
@@ -0,0 +1,224 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <numeric>
+#include <random>
+#include <regex>
+#include <vector>
+
+#include "lm_encoding.hpp"
+#include "openvino/genai/perf_metrics.hpp"
+
+
+namespace ov {
+namespace genai {
+
+// void reset_all_inputs_to_empty_tensors(ov::InferRequest& request) {
+//     request.set_tensor("input_ids", ov::Tensor(ov::element::i64, {0, 0}));
+//     request.set_tensor("beam_idx", ov::Tensor(ov::element::i32, {0}));
+//     if (request.get_compiled_model().inputs().size() == 4)
+//         request.set_tensor("position_ids", ov::Tensor(ov::element::i64, {0, 0}));
+// }
+
+
+EncodedResults get_lm_encoded_results(
+    ov::InferRequest& m_model_runner,
+    const ov::Tensor& input_ids,
+    ov::Tensor attention_mask,
+    const std::shared_ptr<StreamerBase>& streamer_ptr,
+    Sampler& sampler,
+    std::vector<SequenceGroup::Ptr> sequence_groups,
+    std::optional<ov::Tensor> position_ids,
+    std::optional<ov::InferRequest> m_embedding,
+    std::optional<size_t> hidden_size,
+    std::optional<float> scale_emb
+) {
+    std::vector<GenerationHandle> generations;
+    for (SequenceGroup::Ptr sequence_group : sequence_groups) {
+        generations.push_back(std::make_shared<GenerationHandleImpl>(sequence_group->get_generation_stream(), sequence_group->get_sampling_parameters()));
+    }
+
+    ov::Shape prompts_shape = input_ids.get_shape();
+    const size_t batch_size = prompts_shape[0];
+
+    const size_t prompt_len = prompts_shape[1];
+    const size_t max_new_tokens = sequence_groups.at(0)->get_sampling_parameters().get_max_new_tokens(prompt_len);
+
+    // Initialize results and performance metrics.
+    EncodedResults results;
+    auto& raw_perf_counters = results.perf_metrics.raw_metrics;
+    raw_perf_counters.m_new_token_times.reserve(max_new_tokens);
+    raw_perf_counters.m_batch_sizes.reserve(max_new_tokens);
+    raw_perf_counters.m_token_infer_durations.reserve(max_new_tokens);
+    raw_perf_counters.m_inference_durations = {{ MicroSeconds(0.0f) }};
+
+    // Initialize inputs
+    if (m_embedding.has_value())
+        m_model_runner.set_tensor("inputs_embeds", input_ids);
+    else
+        m_model_runner.set_tensor("input_ids", input_ids);
+
+    m_model_runner.set_tensor("attention_mask", attention_mask);
+    
+    if (position_ids.has_value())
+        m_model_runner.set_tensor("position_ids", *position_ids);
+
+    m_model_runner.get_tensor("beam_idx").set_shape({ batch_size });
+    std::fill_n(m_model_runner.get_tensor("beam_idx").data<int32_t>(), batch_size, 0);
```

Comment:
is it valid for chat mode? I supposed that in case of chat we need to continue from previous beam idx